examples/gaia_agent/v2.py

```diff
@@ -0,0 +1,273 @@
+        assert isinstance(start_step, GaiaQuestion)
+        messages[-1]["content"] = PromptRegistry.facts_survey_update.format(
+            task=start_step.content,
+            given=self.facts_list(facts_step.given_facts),
+            lookup=self.facts_list(facts_step.facts_to_lookup),
+            derive=self.facts_list(facts_step.facts_to_derive),
+            guesses=self.facts_list(facts_step.educated_guesses),
+        )
+        return messages
+
+    def facts_list(self, facts: list) -> str:
+        return "\n".join([f" - {fact.name}\n{fact.description}" for fact in facts])
+
+
+class Replan(ThinkingNode):
+    """
+    Produces a new plan based on the reflection of the failed one.
+    """
+
+    system_prompt: str = PromptRegistry.system_prompt
+    guidance: str = PromptRegistry.replan
+
+    def make_prompt(self, agent: Any, tape: Tape) -> Prompt:
+        plan = PlanView(tape)
+        plan_text = "\n".join([f"{step.number}. {step.name}\n{step.description}" for step in plan.steps.values()])
+        plan_result = last_step(tape, PlanReflection)
+        assert plan_result
+        failure_text = f"Failed step: {plan_result.failed_step_number}\n{plan_result.failed_step_overview}"
+        self.guidance = PromptRegistry.replan.format(plan=plan_text, failure=failure_text)
+        return super().make_prompt(agent, tape)
+
+
+class ProduceAnswer(Formalize):
+    """
+    Produces the final answer out of the final plan reflection.
+    """
+
+    agent_step_cls: Any = GaiaAnswer
+
+    def make_prompt(self, agent: Any, tape: Tape) -> Prompt:
+        start_step = tape[0]
+        assert isinstance(start_step, GaiaQuestion)
+        self.guidance = PromptRegistry.final_answer.format(task=start_step.content)
+        return super().make_prompt(agent, tape)
+
+
+class Act(MonoNode):
+    """
+    Act or return to the upper level agent if finished.
+    """
+
+    system_prompt: str = PromptRegistry.system_prompt
+    agent_step_cls: Any = ExecutorStep
+
+    def generate_steps(self, agent: Any, tape: Tape, llm_stream: LLMStream):
+        for step in super().generate_steps(agent, tape, llm_stream):
+            yield step
+            if isinstance(step, FinishSubtask):
+                yield Respond(copy_output=True)
+                break
+
+
+class GaiaPlanner(Agent):
+    @classmethod
+    def create(
+        cls,
+        llm: LLM,
+    ):
+        subagents = [GaiaManager.create(llm)]
+        nodes = (
+            ThinkingNode(
+                name="facts_survey",
+                system_prompt=PromptRegistry.facts_survey_v2_system,
+                guidance=PromptRegistry.facts_survey_v2,
+            ),
+            Formalize(name="formalize_survey", agent_step_cls=ListOfFactsThoughtV2),
+            ThinkingNode(name="plan", system_prompt=PromptRegistry.system_prompt, guidance=PromptRegistry.plan_v2),
+            Formalize(name="formalize_plan", agent_step_cls=PlanThoughtV2, next_agent="GaiaManager"),
```

Comment:
can we move the next agent call to a separate prompt-less node? it's kinda awkward to add `next_agent` field to all Pydantic models

Reply:
I had the same idea and changed that, yes